### ISSUE TITLE
transforms_graph error when "remove_nodes(op=Identity, op=CheckNumerics)" and "fold_old_batch_norms" are used together

### ISSUE THREAD

**Issue by HwMohanLiu** [tensorflow/tensorflow#19530]
### System information
- **Have I written custom code (as opposed to using a stock example script provided in TensorFlow)**: No
- **OS Platform and Distribution (e.g., Linux Ubuntu 16.04)**: Ubuntu 16.04
- **TensorFlow installed from (source or binary)**: source
- **TensorFlow version (use command below)**: r1.8
- **Python version**: 2.7
- **Bazel version (if compiling from source)**: 0.11.0
- **GCC/Compiler version (if compiling from source)**: 4.9
- **CUDA/cuDNN version**: No
- **GPU model and memory**: No
- **Exact command to reproduce**: ` bazel build tensorflow/tools/graph_transforms:transform_graph`

### Describe the problem
I followed the proposed commands for optimizing the model. However, "remove_node(op=Identity, op=CheckNumerics)" and "fold_old_batch_norms" cannot be used together.

### Source code / logs
```
bazel-bin/tensorflow/tools/graph_transforms/transform_graph \
--in_graph=resnet-18_train.pb \
--out_graph=opt_resnet-18_train.pb \
--inputs='input_x' \
--outputs='output/BiasAdd' \
--transforms='
strip_unused_nodes(type=float, shape="1,224,224,3")
remove_nodes(op=Identity, op=CheckNumerics)
fold_old_batch_norms
fold_batch_norms'
```
Error:
> 2018-05-24 20:53:06.184534: I tensorflow/tools/graph_transforms/transform_graph.cc:264] Applying strip_unused_nodes
2018-05-24 20:53:06.259019: I tensorflow/tools/graph_transforms/transform_graph.cc:264] Applying remove_nodes
2018-05-24 20:53:06.486037: I tensorflow/tools/graph_transforms/transform_graph.cc:264] Applying fold_old_batch_norms
2018-05-24 20:53:06.513047: E tensorflow/tools/graph_transforms/transform_graph.cc:210] Beta input to batch norm has bad shape: [64]

However, if "remove_node(op=Identity, op=CheckNumerics)" is removed
```
bazel-bin/tensorflow/tools/graph_transforms/transform_graph \
--in_graph=resnet-18_train.pb \
--out_graph=opt_resnet-18_train.pb \
--inputs='input_x' \
--outputs='output/BiasAdd' \
--transforms='
strip_unused_nodes(type=float, shape="1,224,224,3")
fold_old_batch_norms
fold_batch_norms'
```
It then works...

**Comment by tatianashp** [comment]
This is not surprising. "remove_node(op=Identity, op=CheckNumerics)" optimization breaks the model that contain control flow operations, such as tf.cond, tf.map_fn, and tf.while. Graph Transform Tool [README](https://github.com/tensorflow/tensorflow/blob/r1.8/tensorflow/tools/graph_transforms/README.md#using-the-graph-transform-tool) mentions this.

**Comment by HwMohanLiu** [comment]
@tatianashp thanks a lot. It seems that we do not have any of control flow ops used in the model. According to the risk of using this dangerous option, it will be more meaningful to remove this option from the example? Single-input and single-output ops may often be used and there is no warning about removing them.

**Comment by tatianashp** [comment]
@raingo,  @petewarden - Any thoughts here? Is this a bug or expected behavior for a graph without control flow?

**Comment by tensorflowbutler** [comment]
As this issue has invited community support, please remove the assignee. Otherwise, remove the `community support` label. Thank you.

**Comment by tensorflowbutler** [comment]
As this issue has invited community support, please remove the assignee. Otherwise, remove the `community support` label. Thank you.

**Comment by tensorflowbutler** [comment]
As this issue has invited community support, please remove the assignee. Otherwise, remove the `community support` label. Thank you.

**Comment by tensorflowbutler** [comment]
As this issue has invited community support, please remove the assignee. Otherwise, remove the `community support` label. Thank you.

**Comment by tensorflowbutler** [comment]
As this issue has invited community support, please remove the assignee. Otherwise, remove the `community support` label. Thank you.

**Comment by tensorflowbutler** [comment]
As this issue has invited community support, please remove the assignee. Otherwise, remove the `community support` label. Thank you.

**Comment by tensorflowbutler** [comment]
As this issue has invited community support, please remove the assignee. Otherwise, remove the `community support` label. Thank you.

**Comment by tensorflowbutler** [comment]
As this issue has invited community support, please remove the assignee. Otherwise, remove the `community support` label. Thank you.

**Comment by tensorflowbutler** [comment]
As this issue has invited community support, please remove the assignee. Otherwise, remove the `community support` label. Thank you.

**Comment by tensorflowbutler** [comment]
As this issue has invited community support, please remove the assignee. Otherwise, remove the `community support` label. Thank you.

**Comment by tensorflowbutler** [comment]
As this issue has invited community support, please remove the assignee. Otherwise, remove the `community support` label. Thank you.

**Comment by tensorflowbutler** [comment]
As this issue has invited community support, please remove the assignee. Otherwise, remove the `community support` label. Thank you.

**Comment by IvyGongoogle** [comment]
@HwMohanLiu  when I use 
```
--transforms='
remove_nodes(op=Identity, op=CheckNumerics)
fold_constants(ignore_errors=true)
fold_batch_norms
fold_old_batch_norms
strip_unused_nodes
sort_by_execution_order
'
```
get the same error:

```
2019-01-28 11:17:22.671235: I tensorflow/tools/graph_transforms/transform_graph.cc:318] Applying remove_nodes
2019-01-28 11:17:22.732271: I tensorflow/tools/graph_transforms/transform_graph.cc:318] Applying fold_constants
2019-01-28 11:17:22.783415: I tensorflow/tools/graph_transforms/transform_graph.cc:318] Applying fold_batch_norms
2019-01-28 11:17:22.805463: I tensorflow/tools/graph_transforms/transform_graph.cc:318] Applying fold_old_batch_norms
2019-01-28 11:17:22.825554: E tensorflow/tools/graph_transforms/transform_graph.cc:264] Beta input to batch norm has bad shape: [64]
```
but why do not you use `fold_constants(ignore_errors=true)` before using `fold_batch_norms` and 
`fold_old_batch_norms`?  And after `remove_node(op=Identity, op=CheckNumerics)` is removed as you said,  I still get the same error.

**Comment by HwMohanLiu** [comment]
@IvyGongoogle As tatianashp mentioned, the option "remove_nodes" could break the control flow options in the model. I did not manually  add any control flow ops which tatianashp mentioned. However, it failed. Then, I removed the option "remove_nodes", and the model works well and can be correctly converted into TF-Lite model without BN problems.

**Comment by IvyGongoogle** [comment]
@HwMohanLiu  there is also no control flow ops in my graph. And I doubt you fold_batch_norms successfully  or not.

**Comment by HwMohanLiu** [comment]
@IvyGongoogle If you used FusedBatchNorm (for example tf.layers.batch_normalization), it should be fold_old_batch_norms in my opinion. Although you will still see the node FusedBatchNorm after transform, the content is already changed. There is only a BiasAdd in it. Also you can find the structure mapping definitions for BN in the corresponding source codes in the module https://github.com/tensorflow/tensorflow/tree/master/tensorflow/tools/graph_transforms

**Comment by IvyGongoogle** [comment]
@HwMohanLiu After using `fold_old_batch_norms`, we get a `BiasAdd` for  `FusedBatchNorm`, but there is no `biases` op after `weights` op. But there is a `Conv2D_bn_offset` op,  is this an op with same meaning of `biases`?

**Comment by HwMohanLiu** [comment]
@IvyGongoogle You can determine it by tracing the input ops of the BiasAdd node until finding a const or ref op with the size of the corresponding bias.

**Comment by IvyGongoogle** [comment]
@HwMohanLiu when I trace the input ops of the `BiasAdd`, I find the `Conv2D_bn_offset` op is indeed an op with same meaning of `biases`. Thank you very much. 
But I have another question: the result of infering this merged-BN  pb is slightly different with the original pb, about several numbers.

**Comment by HwMohanLiu** [comment]
@IvyGongoogle If the original pb is used for validation before folding BN, you may have to keep the input batch size as training to get a similar accuracy. After folding, you can input the images one by one. There could be very slight difference between them. However, the difference can be ignored since the gap is normally after four decimals according to my experience.

**Comment by ymodak** [comment]
Graph transform tool is [deprecated](https://github.com/tensorflow/community/blob/master/rfcs/20180907-contrib-sunset.md#list-of-projects) with TF 2.X versions and replaced by [Grappler](https://www.tensorflow.org/guide/graph_optimization).
